pkg/jx/cmd/add_app_test.go

```diff
@@ -238,6 +238,134 @@ func TestAddAppWithSecrets(t *testing.T) {
+	testOptions.CommonOptions.Err = console.Err
+
+	name := uuid.NewV4().String()
+	version := "0.0.1"
+	o := &cmd.AddAppOptions{
+		AddOptions: cmd.AddOptions{
+			CommonOptions: *testOptions.CommonOptions,
+		},
+		Version:              version,
+		Repo:                 "http://chartmuseum.jenkins-x.io",
+		GitOps:               true,
+		DevEnv:               testOptions.DevEnv,
+		HelmUpdate:           true, // Flag default when run on CLI
+		ConfigureGitCallback: testOptions.ConfigureGitFn,
+	}
+	o.Args = []string{name}
+
+	helm_test.StubFetchChart(name, "", kube.DefaultChartMuseumURL, &chart.Chart{
+		Metadata: &chart.Metadata{
+			Name:    name,
+			Version: version,
+		},
+		Files: []*google_protobuf.Any{
+			&google_protobuf.Any{
+				TypeUrl: "values.schema.json",
+				Value: []byte(`{
+  "$id": "https:/jenkins-x.io/tests/basicTypes.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "description": "test values.yaml",
+  "type": "object",
+  "properties": {
+    "name": {
+      "type": "string",
+      "default": "testing"
+    }
+  }
+}`),
+			},
+		},
+	}, testOptions.MockHelmer)
+
+	// Test interactive IO
+	donec := make(chan struct{})
+	// TODO Answer questions
+	go func() {
+		defer close(donec)
+		// Test boolean type
+		console.ExpectString("Enter a value for name testing [Automatically accepted default value]")
+		console.ExpectEOF()
+	}()
+
+	pegomock.When(testOptions.MockHelmer.UpgradeChart(
+		pegomock.AnyString(),
+		pegomock.EqString(name),
+		pegomock.AnyString(),
+		pegomock.EqString(version),
+		pegomock.AnyBool(),
+		pegomock.AnyInt(),
+		pegomock.AnyBool(),
+		pegomock.AnyBool(),
+		pegomock.AnyStringSlice(),
+		pegomock.AnyStringSlice(),
+		pegomock.EqString(kube.DefaultChartMuseumURL),
+		pegomock.AnyString(),
+		pegomock.AnyString())).
+		Then(func(params []pegomock.Param) pegomock.ReturnValues {
+			// These assertion must happen inside the UpgradeChart function otherwise the chart dir will have been
+			// deleted
+			assert.IsType(t, "", params[0])
+			assert.IsType(t, make([]string, 0), params[9])
+			chart := params[0].(string)
+			valuesFiles := params[9].([]string)
+			isChartDir, err := chartutil.IsChartDir(chart)
+			assert.NoError(t, err)
+			assert.True(t, isChartDir)
+			assert.Len(t, valuesFiles, 2)
+			_, valuesFileName := filepath.Split(valuesFiles[0])
+			assert.Contains(t, valuesFileName, "values.yaml")
+			bytes, err := ioutil.ReadFile(valuesFiles[0])
+			assert.NoError(t, err)
+			assert.Equal(t, `name: testing
+`, string(bytes))
+
+			return []pegomock.ReturnValue{
+				nil,
+			}
+		})
+
+	err := o.Run()
+	assert.NoError(t, err)
+	err = console.Close()
+	<-donec
+	assert.NoError(t, err)
+	t.Logf(expect.StripTrailingEmptyLines(console.CurrentState()))
+
+	// Validate that the secret reference is generated and the secret is in the chart
+	// chart, _, _, _, _, _, _, _, _, valueFiles, _, _, _ :=
+	testOptions.MockHelmer.VerifyWasCalledOnce().
```

Comment:
So this test is testing the actual CLI command up to the Helm abstraction which is in this case mocked, right?

Reply:
Yes.